### PR TITLE
D3ASIM-3556: Export market structures to d3a-interface

### PR DESCRIPTION
**Pull request by Hassan754**

### CODE REVIEW

File: tests/test_dataclasses.py
Comment (faizan2590) [inline tests/test_dataclasses.py:102]:
```suggestion
            "id": str(uuid.uuid4()),
```

File: d3a_interface/data_classes.py
Comment (faizan2590) [inline d3a_interface/data_classes.py:92]: Why time only needs to be injected for Offer and not for Bid ?

File: d3a_interface/data_classes.py
Comment (faizan2590) [inline d3a_interface/data_classes.py:102]: These parameters should not have a default None and must always be provided

File: d3a_interface/data_classes.py
Comment (faizan2590) [inline d3a_interface/data_classes.py:176]: Where do you need it?

File: tests/test_dataclasses.py
Comment (faizan2590) [inline tests/test_dataclasses.py:217]: These should be a non None

File: d3a_interface/data_classes.py
Comment (faizan2590) [inline d3a_interface/data_classes.py:144]: rounding is done in many places with same decimal places, better to replace it with a common rounding constant variable

File: d3a_interface/data_classes.py
Comment (faizan2590) [inline d3a_interface/data_classes.py:118]: Although its more like a migration from d3a to d3a-interface BUT still how about using f-string now

**Review by faizan2590 — COMMENTED** [review]
Left some comments, will give another look once addressed/answered.

File: d3a_interface/data_classes.py
Comment (fievelk) [inline d3a_interface/data_classes.py:29]: Nit: I believe this should be renamed.

File: d3a_interface/data_classes.py
Comment (Hassan754) [reply d3a_interface/data_classes.py:102]: I'm not sure, there are dozens of calls to Bid/ Offer instantiation (especially in tests) and they were always omitted
I have checked and it looks like in the production code, these are always sent
Basically I agree that it makes sense to not allow these to be None, however this will require changes to all unit tests, do you think these can be postponed to another time?

File: d3a_interface/data_classes.py
Comment (Hassan754) [reply d3a_interface/data_classes.py:238]: From the official documentation of Optional:

> Optional[X] is equivalent to Union[X, None].
So this is not relevant here

File: d3a_interface/data_classes.py
Comment (Hassan754) [reply d3a_interface/data_classes.py:176]: This can be useful when trying to set an offer/bid or any other market structure as a dictionary key, it will know how to hash the object

File: d3a_interface/data_classes.py
Comment (Hassan754) [reply d3a_interface/data_classes.py:144]: They are used 3 times with a value of 4 for the energy_rate, however I do not think that we should add a constant for them (at least not a global constant ) we may add a constant at the top of the file but The way I see it, this is not critical and would restrict us from having different decimal places for the existing different round calls
Anyway I could add the constant if you believe it's worth it

File: d3a_interface/data_classes.py
Comment (spyrostz) [reply d3a_interface/data_classes.py:92]: Should be for both the same. However keep in mind that this is probably the time in string format, json.loads is not able to deserialize datetimes.

File: d3a_interface/data_classes.py
Comment (spyrostz) [reply d3a_interface/data_classes.py:176]: Please move it to the base class, this applies to all child classes.

**Review by spyrostz — DISMISSED** [review]
LGTM once comments are resolved. I was thinking that it could be worth it to make the base class an ABC in order to know which methods need to be implemented if we plan to add another similar class in the future, however please do not do it because 1. I do not foresee it  2. will unnecessary complicate the existing classes' implementation  3. this PR is urgent

File: d3a_interface/data_classes.py
Comment (Hassan754) [reply d3a_interface/data_classes.py:176]: @spyrostz for some reason when I implement __hash__() in the parent class, the children are not inheriting it, I'm still investigating why 
But it's probably due to the fact the that we're overriding the __eq__ in children classes https://stackoverflow.com/a/2134487

Comment (spyrostz) [reply d3a_interface/data_classes.py:176]: I see, in this case LGTM

**Comment by Hassan754** [comment]
> LGTM once comments are resolved. I was thinking that it could be worth it to make the base class an ABC in order to know which methods need to be implemented if we plan to add another similar class in the future, however please do not do it because 1. I do not foresee it 2. will unnecessary complicate the existing classes' implementation 3. this PR is urgent

Thanks a lot!
There is always extra room for improvements 😫

File: d3a_interface/data_classes.py
Comment (Hassan754) [reply d3a_interface/data_classes.py]: @faizan2590 seems great, done

**Review by spyrostz — APPROVED** [review]
LGTM

File: d3a_interface/data_classes.py
Comment (faizan2590) [reply d3a_interface/data_classes.py:239]: Sorry, didn'T quite get you, here `already_tracked ` isn't a union with None. Can you please explain a bit more.

File: d3a_interface/data_classes.py
Comment (faizan2590) [reply d3a_interface/data_classes.py:103]: Okay, please make a ticket to be in radar that this is something to be dealt with.

File: d3a_interface/data_classes.py
Comment (Hassan754) [reply d3a_interface/data_classes.py:239]: the already_tracked's default value is False, so it's either True or False -> bool
Optional can be added if it can also be None, but it doesn't

File: d3a_interface/data_classes.py
Comment (Hassan754) [reply d3a_interface/data_classes.py:103]: I'll keep the same ticket and move it to SFD after this is merged, and I'll update its description to include more refactorings in the future (I already have some in mind)

File: d3a_interface/data_classes.py
Comment (fievelk) [reply d3a_interface/data_classes.py:239]: @faizan2590 maybe the misunderstanding is related to the name "Optional". In type hints, "Optional" doesn't mean that the parameter can be omitted when calling the function. It means that the value of the parameter can be `None` in some cases. In this situation (as Hassan explained) the `already_tracked` parameter can't be `None`, so using `Optional` would not be correct.

File: d3a_interface/data_classes.py
Comment (faizan2590) [reply d3a_interface/data_classes.py:145]: We have a similar [round_floats_for_ui](https://github.com/gridsingularity/d3a-interface/blob/master/d3a_interface/utils.py#L302). So, could use it here too, off-course its rounding upto 3 decimal points.

Anyways non-blocking

File: tests/test_dataclasses.py
Comment (faizan2590) [reply tests/test_dataclasses.py:219]: https://github.com/gridsingularity/d3a-interface/pull/253/files?file-filters%5B%5D=.py&hide-deleted-files=true#r708031409

File: d3a_interface/data_classes.py
Comment (faizan2590) [reply d3a_interface/data_classes.py:239]: Thank You both for explanation, got it now

File: d3a_interface/data_classes.py
Comment (faizan2590) [reply d3a_interface/data_classes.py:103]: > I'll keep the same ticket and move it to SFD after this is merged, I'll update its description to include more refactorings in the future (I already have some in mind)

If the description of ticket needs to be updated at the time of review, it means thing weren't planned thought of before starting and its something additional, so I would vote to move this ticket to done once this PR is merged, and create a follow-up ticket targeting further enhancements.

Tagging @fievelk @spyrostz in case they don't agree with me .

**Review by faizan2590 — APPROVED** [review]
LGTM

File: d3a_interface/data_classes.py
Comment (Hassan754) [reply d3a_interface/data_classes.py:103]: I'm okay with creating a new ticket, but it will also be a broad : refactor market structures
THe original tickets did not explicitly describe the needed changes indeed, it's a `refactoring` ticket so I resort to my instincts on the go while refactoring, but the majority of changes were mentioned in an older ticket (mostly comments from Pierpaolo) + I added some more changes and left some for the future

Comment (fievelk) [reply d3a_interface/data_classes.py:103]: Since you want to first merge this PR, and then create a new one to address this specific comment, I would say that we can safely open a new ticket. It will also help to track the changes with the commit messages' prefixes.

**Pull request merged**
